### ISSUE TITLE
UnboundLocalError when passing "empty" generator as filenames

### ISSUE THREAD

**Issue by gerritholl** [pytroll/satpy#956]
**Describe the bug**
Passing a generator that yields zero results (i.e. an "empty" generator) will cause a `UnfoundLocalError` in the `satpy.Scene` constructor.  One source of such an empty generator can be the `pathlib.Path.glob` method, which yields its results as a generator object.

**To Reproduce**

```python
import satpy
import pathlib
sc = satpy.Scene(
        sensor="abi",
        reader=["abi_l1b"],
        filenames=pathlib.Path("/not").glob("found"))

```

**Expected behavior**
I expect that this has the same result as passing an empty list, and that an exception is raised saying `ValueError: No supported files found`.

**Actual results**

```
Traceback (most recent call last):
  File "/tmp/mwe16.py", line 6, in <module>
    filenames=pathlib.Path("/not").glob("found"))
  File "/media/nas/x21324/miniconda3/envs/py37e/lib/python3.7/site-packages/satpy/scene.py", line 156, in __init__
    reader_kwargs=reader_kwargs)
  File "/media/nas/x21324/miniconda3/envs/py37e/lib/python3.7/site-packages/satpy/scene.py", line 201, in create_reader_instances
    ppp_config_dir=self.ppp_config_dir)
  File "/media/nas/x21324/miniconda3/envs/py37e/lib/python3.7/site-packages/satpy/readers/__init__.py", line 724, in load_readers
    if loadables:
UnboundLocalError: local variable 'loadables' referenced before assignment
```

**Environment Info:**
 - OS: Linux
 - Satpy Version: 0.17.2.dev173+ge958d7d9

**Comment by djhoese** [comment]
I really thought I fixed this error.

Regardless, I'm not sure a generator is currently supported for this argument but I guess there is no reason it couldn't be.

**Comment by djhoese** [comment]
I reviewed the relevant code for this and have an answer for why this is happening. Generators (or anything that is something other than `None`, a list, a tuple, or a dict) are not technically supported for the `filenames` keyword argument. This is confusing since the docstring (that I wrote I think) says "iterable or dict", but it should actually be `None` (default), `list`, `tuple`, or `dict`. Here I think I poorly chose iterable to mean list/tuple.

Iterators/generators may have been supported in the past but as bugs and errors were fixed over time the if/elif/else block that handles the reader/filenames arguments started to depend on `filenames` be lists or tuples. There are if statements like `if not filenames` to check if the user provided an empty list and provide a specific error message.

Possible solutions:

1. Do `filenames = list(filenames)` as soon as possible.
2. Rework the if statements to allow for filenames to be an iterable/generator, even if that means doing `if hasattr(filenames, '__next__')`.
3. Add an initial, possibly redundant, if statement right before using the filenames to convert it from a generator (we need it to be a `set` object).

Regardless, this `loadables` stuff should be fixed so it produces a more logical error (ex. "No usable files found").
